fdbserver/networktest.actor.cpp

```diff
@@ -233,3 +236,226 @@ ACTOR Future<Void> networkTestClient( std:: string testServers ) {
+		}
+
+		return Void();
+	}
+
+	ACTOR static Future<Void> outgoing(P2PNetworkTest *self) {
+		loop {
+			wait(delay(0, TaskPriority::WriteSocket));
+			state NetworkAddress remote = self->randomRemote();
+
+			try {
+				state Reference<IConnection> conn = wait(INetworkConnections::net()->connect(remote));
+				wait(conn->connectHandshake());
+				//printf("Connected to %s\n", remote.toString().c_str());
+				++self->sessionsOut;
+				wait(doSession(self, conn));
+			} catch(Error &e) {
+				printf("outgoing: error %s on remote %s\n", e.what(), remote.toString().c_str());
+			}
+		}
+	}
+
+	ACTOR static Future<Void> incoming(P2PNetworkTest *self, Reference<IListener> listener) {
+		state ActorCollection sessions(false);
+
+		loop {
+			wait(delay(0, TaskPriority::AcceptSocket));
+
+			try {
+				state Reference<IConnection> conn = wait(listener->accept());
+				wait(conn->acceptHandshake());
+				//printf("Connected from %s\n", conn->getPeerAddress().toString().c_str());
+				++self->sessionsIn;
+				sessions.add(doSession(self, conn));
+			} catch(Error &e) {
+				printf("incoming: error %s on listener %s\n", e.what(), listener->getListenAddress().toString().c_str());
+			}
+		}
+	}
+
+	ACTOR static Future<Void> run_impl(P2PNetworkTest *self) {
+		state ActorCollection actors(false);
+		g_network->initTLS();
+
+		self->startTime = now();
+
+		printf("%d listeners, %d remotes, %d outgoing connections\n", self->listeners.size(), self->remotes.size(), self->connectionsOut);
+		printf("Message size %d to %d bytes\n", self->msgBytes.min, self->msgBytes.max);
+		printf("Post exchange delay %f to %f seconds\n", self->idleMilliseconds.min / 1e3, self->idleMilliseconds.max / 1e3);
+		printf("Send/Recv size %d bytes\n", self->sendRecvSize);
+
+		for(auto n : self->remotes) {
+			printf("Remote: %s\n", n.toString().c_str());
+		}
+
+		for(auto el : self->listeners) {
+			printf("Listener: %s\n", el->getListenAddress().toString().c_str());
+			actors.add(incoming(self, el));
+		}
+
+		if(!self->remotes.empty()) {
+			for(int i = 0; i < self->connectionsOut; ++i) {
+				actors.add(outgoing(self));
+			}
+		}
+
+		loop {
+			wait(delay(1.0));
+			printf("%s\n", self->statsString().c_str());
+		}
+	}
+
+	Future<Void> run() {
+		return run_impl(this);
+	}
+
+};
+
+int getEnvInt(const char *name, int defaultValue = 0) {
+	const char *val = getenv(name);
+	return val != nullptr ? atol(val) : defaultValue;
+}
+
+std::string getEnvStr(const char *name, std::string defaultValue = "") {
+	const char *val = getenv(name);
+	return val != nullptr ? val : defaultValue;
+}
+
+// TODO: Remove this hacky thing and make a "networkp2ptest" role in fdbserver
+TEST_CASE("!p2ptest") {
+	state P2PNetworkTest p2p(
+		getEnvStr("listenerAddresses", "127.0.0.1:5000,127.0.0.1:5001:tls"),
+		getEnvStr("remoteAddresses", "127.0.0.1:5000,127.0.0.1:5001:tls"),
+		getEnvInt("connectionsOut", 2),
```

Comment:
This only opens two connections per test?  I'd expect we'd need 100's-1000's of concurrent handshakes to find any races int the code.

Reply:
This was just a way to get the test running, and a very hacky one, hence the TODO above it to make this an `fdbserver` role.  By minimizing or maximizing different things you can stress and measure different things.  I don't have a strong opinion about the default parameters.